### PR TITLE
Add mock responses for OffsetManager testing

### PR DESCRIPTION
**Pull request by horkhe**
This PR brings mock response builders and respective additions to response structs that are used by offset manager.

### CODE REVIEW

File: consumer_metadata_response.go [hunk 1]
```diff
@@ -20,10 +20,14 @@ func (r *ConsumerMetadataResponse) decode(pd packetDecoder) (err error) {
 	}
 	r.Err = KError(tmp)
 
-	r.Coordinator = new(Broker)
-	if err := r.Coordinator.decode(pd); err != nil {
+	coordinator := new(Broker)
+	if err := coordinator.decode(pd); err != nil {
 		return err
 	}
+	if coordinator.addr == ":0" {
```
Comment (eapache) [inline consumer_metadata_response.go]: Why this change? Does Kafka return an address of ":0" in special cases?

File: mockresponses_test.go [hunk 2]
```diff
@@ -230,3 +230,152 @@ func (mfr *mockFetchResponse) getHighWaterMark(topic string, partition int32) in
...
+	res := &OffsetFetchResponse{}
+	for topic, partitions := range mr.offsets[group] {
+		for partition, block := range partitions {
+			res.AddBlock(topic, partition, block)
+		}
+	}
+	return res
+}
+
+//func (mr *mockOffsetFetchResponse) getOffset(group, topic string, partition int32) *OffsetFetchResponseBlock {
```
Comment (eapache) [inline mockresponses_test.go]: why is this commented out?

File: offset_commit_response.go [hunk 3]
```diff
@@ -4,9 +4,40 @@ type OffsetCommitResponse struct {
...
+			pe.putInt16(int16(kerror))
+		}
+	}
+	return nil
+}
+
 func (r *OffsetCommitResponse) decode(pd packetDecoder) (err error) {
 	numTopics, err := pd.getArrayLength()
-	if err != nil {
+	if err != nil || numTopics == 0 {
```
Comment (eapache) [inline offset_commit_response.go]: this extra conditional seems unnecessary, it would just fall through the loop without it?

**Comment by eapache** [comment]
A few minor questions but generally looks good.

cc @aaronkavlie-wf you may want to coordinate with @horkhe since it looks like you're both interested in pushing the offset manager forward

File: offset_commit_response.go (hunk 3 above)
Comment (horkhe) [inline offset_commit_response.go]: It is necessary to make encode/decode symmetric. When a reply that contains no topics encoded it used to contain an empty slice of blocks, but a response object that contains no topics had `nil` slice of blocks instead, so they did not compare equal.

File: mockresponses_test.go (hunk 2 above)
Comment (horkhe) [inline mockresponses_test.go]: It is a remnant of the previous iteration of the implementation, I will remove that.

File: consumer_metadata_response.go (hunk 1 above)
Comment (horkhe) [inline consumer_metadata_response.go]: It is necessary to make encode/decode symmetric. When a response that contains zero coordinator info is encoded it used to contain a zero instance of broker, but a response object that contains no broker had a `nil` broker reference.

**Comment by aaronkavlie-wf** [comment]
Looks like this should take care of some of the hangups I ran into with testing, at least (missing decode/encode methods).

@horkhe have you done any work on tests for the offset manager PR?

**Comment by horkhe** [comment]
@aaronkavlie-wf sorry, man, but I have written [our own](https://github.com/mailgun/sarama/blob/maxim/develop/offset_manager.go) offset manager, and it is tested. With this PR I am sharing some code from that effort with the upstream repo.

**Comment by aaronkavlie-wf** [comment]
@horkhe hmm, why did you decide to write your own?

(Perhaps we should say "fork" as I see that it's clearly based on [the outstanding PR](https://github.com/Shopify/sarama/pull/461)

**Comment by horkhe** [comment]
@aaronkavlie-wf when we wrote it the OffsetManager in the outstanding PR had different interface and a number of issues. You can find the list in one of my comments to the PR. Besides it our implementation the OffsetManager guarantees to commit pending offsets on close, whereas the OffsetManger in the outstanding PR drops them.

**Comment by eapache** [comment]
code :+1: now

**Comment by aaronkavlie-wf** [comment]
@horkhe right, but why not submit the changes as a PR to Sarama?

We have a need for offset management as well, but I'm looking to get it into Sarama to avoid the need to maintain a separate fork.
